cmd/clusterctl/test/e2e/delete_test.go

```diff
@@ -0,0 +1,131 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterctlclient "sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
+)
+
+var _ = Describe("clusterctl delete", func() {
+	var (
+		mgmtInfo      testMgmtClusterInfo
+		deleteOptions clusterctlclient.DeleteOptions
+	)
+
+	BeforeEach(func() {
+		var err error
+		// Mgmt cluster info object
+		mgmtInfo = testMgmtClusterInfo{
+			clusterctlConfigFile:    clusterctlConfigFile,
+			coreProvider:            "cluster-api:v0.3.0",
+			bootstrapProviders:      []string{"kubeadm-bootstrap:v0.3.0"},
+			controlPlaneProviders:   []string{"kubeadm-control-plane:v0.3.0"},
+			infrastructureProviders: []string{"docker:v0.3.0"},
+		}
+		// Create the mgmt cluster and client
+		mgmtInfo.mgmtCluster, err = CreateKindCluster(kindConfigFile)
+		Expect(err).ToNot(HaveOccurred())
+		mgmtInfo.mgmtClient, err = mgmtInfo.mgmtCluster.GetClient()
+		Expect(err).NotTo(HaveOccurred())
+
+		initTestMgmtCluster(ctx, mgmtInfo)
+
+	}, setupTimeout)
+
+	JustBeforeEach(func() {
+		c, err := clusterctlclient.New(mgmtInfo.clusterctlConfigFile)
+		Expect(err).ToNot(HaveOccurred())
+		err = c.Delete(deleteOptions)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		fmt.Fprintf(GinkgoWriter, "Tearing down kind clusters\n")
+		mgmtInfo.mgmtCluster.Teardown(ctx)
+	})
+
+	Context("deletes the infra provider", func() {
+		BeforeEach(func() {
+			deleteOptions = clusterctlclient.DeleteOptions{
+				Kubeconfig: mgmtInfo.mgmtCluster.KubeconfigPath,
+				Providers:  []string{"docker"},
+			}
+		})
+		It("should delete of all infra provider components except the hosting namespace and the CRDs.", func() {
+			Eventually(
+				func() error {
+					err := mgmtInfo.mgmtClient.Get(ctx, client.ObjectKey{Namespace: "capd-system", Name: "capd-controller-manager"}, &appsv1.Deployment{})
+					if err == nil || !apierrors.IsNotFound(err) {
+						// deployment still exists or some other error other than not found occured.
+						return fmt.Errorf("%v", err)
+					}
+					return nil
+				}, 3*time.Minute, 5*time.Second,
+			).ShouldNot(HaveOccurred())
+		})
+	})
+	Context("deletes everything", func() {
+		BeforeEach(func() {
+			deleteOptions = clusterctlclient.DeleteOptions{
+				Kubeconfig:           mgmtInfo.mgmtCluster.KubeconfigPath,
+				ForceDeleteNamespace: true,
+				ForceDeleteCRD:       true,
+				Providers:            []string{},
+			}
+		})
+		It("should reset the management cluster to its original state", func() {
+			Eventually(
+				func() error {
+					// TODO: check all components are deleted.
+					err := mgmtInfo.mgmtClient.Get(ctx, client.ObjectKey{Namespace: "capd-system", Name: "capd-controller-manager"}, &appsv1.Deployment{})
+					if err == nil || !apierrors.IsNotFound(err) {
```

Comment:
should these all be `err != nil && !apierrors.IsNotFound(err)`?

Reply:
Please see above comment.

Reply:
This can be simplified to `!apierrors.IsNotFound(err)` which is a little more clear in intent

I would also recommend adding a message to the error just to explain what happened, such as

```go
    return fmt.Errorf("The object still exists or there was an error retrieving it: %v", err)
```

Reply:
Done.